jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/ExtractTarStep.java

```diff
@@ -0,0 +1,140 @@
+      inputStream.mark(2);
+      int magic = (inputStream.read() & 0xff) | ((inputStream.read() << 8) & 0xff00);
+      return magic == GZIPInputStream.GZIP_MAGIC;
+    }
+  }
+
+  private final Path tarPath;
+  private final Path destination;
+
+  ExtractTarStep(Path tarPath, Path destination) {
+    this.tarPath = tarPath;
+    this.destination = destination;
+  }
+
+  @Override
+  public LocalImage call()
+      throws IOException, LayerCountMismatchException, BadContainerConfigurationFormatException {
+    TarExtractor.extract(tarPath, destination);
+    DockerManifestEntryTemplate loadManifest =
+        new ObjectMapper()
+            .configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true)
+            .readValue(
+                Files.newInputStream(destination.resolve("manifest.json")),
+                DockerManifestEntryTemplate[].class)[0];
+    ContainerConfigurationTemplate configuration =
+        JsonTemplateMapper.readJsonFromFile(
+            destination.resolve(loadManifest.getConfig()), ContainerConfigurationTemplate.class);
+
+    List<String> layerFiles = loadManifest.getLayerFiles();
+    if (configuration.getLayerCount() != layerFiles.size()) {
+      throw new LayerCountMismatchException(
+          "Invalid base image format: manifest contains "
+              + layerFiles.size()
+              + " layers, but container configuration contains "
+              + configuration.getLayerCount()
+              + " layers");
+    }
+
+    // Check the first layer to see if the layers are compressed already. 'docker save' output is
+    // uncompressed, but a jib-built tar has compressed layers.
+    boolean layersAreCompressed =
+        layerFiles.size() > 0 && isGzipped(destination.resolve(layerFiles.get(0)));
+
+    // Process layer blobs
+    // TODO: Optimize; compressing/calculating layer digests is slow
+    List<PreparedLayer> layers = new ArrayList<>();
+    V22ManifestTemplate newManifest = new V22ManifestTemplate();
+    for (int index = 0; index < layerFiles.size(); index++) {
+      Path file = destination.resolve(layerFiles.get(index));
+
+      // Compress layers if necessary and calculate the digest/size
+      Blob blob = layersAreCompressed ? Blobs.from(file) : Blobs.compress(Blobs.from(file));
+      BlobDescriptor blobDescriptor = blob.writeTo(ByteStreams.nullOutputStream());
+
+      // 'manifest' contains the layer files in the same order as the diff ids in 'configuration',
+      // so we don't need to recalculate those.
+      // https://containers.gitbook.io/build-containers-the-hard-way/#docker-load-format
+      CachedLayer layer =
+          CachedLayer.builder()
+              .setLayerBlob(blob)
+              .setLayerDigest(blobDescriptor.getDigest())
+              .setLayerSize(blobDescriptor.getSize())
+              .setLayerDiffId(configuration.getLayerDiffId(index))
+              .build();
+
+      // TODO: Check blob existence on target registry (online mode only)
```

Comment:
To me, the comment is not clear why we need/want to do this for what. (I'm wondering it too.)

Reply:
I'll just remove this, I mixed up what the purpose of the blob existence check was.